lib/util/createDomain.js

```diff
@@ -10,8 +10,12 @@ module.exports = function createDomain(options, listeningApp) {
   const port = options.socket ? 0 : appPort;
   const hostname = options.useLocalIp ? internalIp.v4() : options.host;
 
+  // use explicitly defined public url (prefix with protocol if not explicitly given)
+  if (options.public) {
+    return /^[a-zA-Z]+:\/\//.test(options.public) ? `${options.public}` : `${protocol}://${options.public}`;
```

Comment:
sorry, but this is not what was requested in previous comments

> this is not a good way to check a url for a protocol. please use [`url.parse`](https://nodejs.org/api/url.html#url_url_parse_urlstring_parsequerystring_slashesdenotehost) instead, and examine the result for a protocol.

Reply:
I wrote in the commit comment that `url.parse` is not appropriate here.. I did it with `url.parse` before but the new tests I provided showed it's not working as it should..

> found a bug due to tests: `url.parse` doesn't properly parse a provided `public` option since it counts everything before the port colon as the protocol which is just wrong. I replaced it with a protocol regex.

Reply:
Sorry, but I looked up url.parse documentation again, as well as searching for issues with what I'm experiencing, and there's no easy solution possible by using url.parse.

Why? The `public` parameter can be a non-valid url like "localhost:8080" which is parsed to somethin like this:
```javascript
> url.parse("localhost:8080")
Url {
  protocol: 'localhost:',
  slashes: null,
  auth: null,
  host: '8080',
  port: null,
  hostname: '8080',
  hash: null,
  search: null,
  query: null,
  pathname: null,
  path: null,
  href: 'localhost:8080' }
```
As you can see, the protocol is incorrect and I would need to do more processing to find out what's going on. I think the solution using regex is the better option here.

Also have a look at https://github.com/nodejs/node/issues/5755, where they're discussing the same issue on parsing urls.

Reply:
great documentation of the why. that'll certainly be beneficial in the future